spark-3.1-spanner-lib/src/test/java/com/google/cloud/spark/spanner/ReadIntegrationTestBase.java

```diff
@@ -212,7 +219,196 @@ public void testDataset_limit() {
+    List<Object> gotBsObj = new ArrayList();
+    gotBs.forEach(v -> gotBsObj.add(v));
+    Collections.sort(wantBs, cmp);
+    Collections.sort(gotBsObj, cmp);
+    assertThat(gotBsObj).isEqualTo(wantBs);
+
+    // STRING: NULL.
+    List<String> gotCs = df.select("C").as(Encoders.STRING()).collectAsList();
+    List<String> wantCs = Arrays.asList(((String) null), ((String) null), "😎🚨", "🚨");
+
+    Collections.sort(wantCs, cmp);
+    Collections.sort(gotCs, cmp);
+    assertThat(gotCs).isEqualTo(wantCs);
+
+    // NUMERIC: NULL.
+    List<BigDecimal> gotDs = df.select("D").as(Encoders.DECIMAL()).collectAsList();
+    List<BigDecimal> wantDs =
+        Arrays.asList(null, null, asSparkBigDecimal("55.7"), asSparkBigDecimal("99.37171"));
+    Collections.sort(gotDs, cmp);
+    Collections.sort(wantDs, cmp);
+    assertThat(gotDs).isEqualTo(wantDs);
+
+    // DATE: NULL.
+    List<Date> gotEs = df.select("E").as(Encoders.DATE()).collectAsList();
+    List<Date> wantEs = Arrays.asList(null, null, null, Date.valueOf("2023-12-30"));
+    Collections.sort(gotEs, cmp);
+    Collections.sort(wantEs, cmp);
+    assertThat(gotEs).isEqualTo(wantEs);
+
+    // TIMESTAMP: NULL.
+    List<Timestamp> gotFs = df.select("F").as(Encoders.TIMESTAMP()).collectAsList();
+    List<Timestamp> wantFs =
+        Arrays.asList(null, null, null, asSparkTimestamp("2023-09-23T12:11:09Z"));
+    Collections.sort(gotFs, cmp);
+    Collections.sort(wantFs, cmp);
+    assertThat(gotFs).isEqualTo(wantFs);
+
+    // BOOL: NULL.
+    List<Boolean> gotGs = df.select("G").as(Encoders.BOOLEAN()).sort().collectAsList();
+    List<Boolean> wantGs = Arrays.asList(null, true, false, false);
+    Collections.sort(gotGs, cmp);
+    Collections.sort(wantGs, cmp);
+    assertThat(gotGs).isEqualTo(wantGs);
+
+    // ARRAY<DATE>: NULL.
+    List<Row> allHs = df.select("H").collectAsList();
+    assertEquals(allHs.size(), 4);
+    List<Date> gotHs = new ArrayList();
+
+    for (int i = 0; i < allHs.size(); i++) {
+      Row row = allHs.get(i);
+
+      for (int j = 0; j < row.size(); j++) {
+        Object obj = row.get(j);
+        if (obj == null) {
+          continue;
+        }
+
+        List<Date> dj = row.getList(j);
+        gotHs.addAll(dj);
+      }
+    }
+
+    List<Object> wantHs =
+        Arrays.asList(
+            ((Date) null), ((Date) null), Date.valueOf("2022-10-01"), Date.valueOf("2023-09-22"));
+    List<Object> gotHsObj = new ArrayList();
+    gotHs.forEach(v -> gotHsObj.add(v));
+    Collections.sort(wantHs, cmp);
+    Collections.sort(gotHsObj, cmp);
+    assertThat(gotHsObj).isEqualTo(wantHs);
+
+    // ARRAY<TIMESTAMP>: NULL.
+    List<Row> allIs = df.select("I").collectAsList();
+    assertEquals(allIs.size(), 4);
+    List<Timestamp> gotIs = new ArrayList();
+
+    for (int i = 0; i < allIs.size(); i++) {
+      Row row = allIs.get(i);
+
+      for (int j = 0; j < row.size(); j++) {
+        Object obj = row.get(j);
+        if (obj == null) {
```

Comment:
We'll need to test how many nulls are there.

Reply:
Done, thanks @halio-g!